### ISSUE TITLE
請問4.0release版本的WebRTC推拉流架構以及WebRTC是否不消耗srs connection數量？

### ISSUE THREAD

**Issue by elin72215** [ossrs/srs#2645]
> 注意：不提供以下信息的Issue会被直接删除(Please follow issue template, or we will delete it)

> 注意：咨询和讨论请提交到SRS星球(Please ask question at) http://bbs.ossrs.net

**描述(Description)**

       在下看了wiki但是沒找到srs實現WebRTC推拉流的架構，本來我以為是推流統一轉成RTMP然後拉流再轉成WebRTC，如此在進行直播時就會因為多客戶端拉同一個流時佔用多條連線。
       於是設定了max_connections為3來測試當連線佔滿時會發生什麼情況，因為想順便測試這篇文章https://www.zhihu.com/question/25497090有人說WebRTC只能支持十條以下連線不適合直播，實際上我開了十幾個播放器來播放同一個流，十幾個播放器播放都正常播放，完全不受max_connections限制。
       那麼問題來了，當我走WebRTC推拉流時：
           1.拉流是不受max_connections限制？還是有其他內定的限制？
           2.那麼推流是否有限制呢？
           3.WebRTC拉流規模要到多大時才會發生畫面模糊或Srs已經優化了這個部分？
           4.是否有WebRTC推拉流架構的文件或架構圖呢？

1. SRS版本(Version): `4.0release`
1. SRS的日志如下(Log):
```
[2021-10-01 06:03:10.883][Trace][1][1n885wu3] client identified, type=flash-publish, vhost=srs-origin-0.socs, app=live, stream=999, param=?eip=35.221.153.50, duration=0ms
[2021-10-01 06:03:10.883][Trace][1][1n885wu3] connected stream, tcUrl=rtmp://srs-origin-0.socs/live, pageUrl=, swfUrl=, schema=rtmp, vhost=__defaultVhost__, port=1935, app=live, stream=999, param=?eip=35.221.153.50, args=(obj)
[2021-10-01 06:03:10.883][Trace][1][1n885wu3] source url=/live/999, ip=127.0.0.1, cache=1, is_edge=0, source_id=712p9886/712p9886
[2021-10-01 06:03:10.884][Trace][1][1n885wu3] TCP: before dispose resource(RtmpConn)(0x351cdb0), conns=3, zombies=0, ign=0, inz=0, ind=0
[2021-10-01 06:03:10.884][Error][1][1n885wu3][11] serve error code=1028 : service cycle : rtmp: stream service : rtmp: stream /live/999 is busy
thread [1][1n885wu3]: do_cycle() [src/app/srs_app_rtmp_conn.cpp:217][errno=11]
thread [1][1n885wu3]: service_cycle() [src/app/srs_app_rtmp_conn.cpp:414][errno=11]
thread [1][1n885wu3]: acquire_publish() [src/app/srs_app_rtmp_conn.cpp:949][errno=11](Resource temporarily unavailable)
[2021-10-01 06:03:10.884][Trace][1][j11ma521] TCP: clear zombies=1 resources, conns=3, removing=0, unsubs=0
[2021-10-01 06:03:10.884][Trace][1][1n885wu3] TCP: disposing #0 resource(RtmpConn)(0x351cdb0), conns=3, disposing=1, zombies=0
[2021-10-01 06:03:11.118][Trace][1][712p9886] 34B video sh,  codec(7, profile=Baseline, level=2.1, 320x240, 0kbps, 0.0fps, 0.0s)
[2021-10-01 06:03:11.233][Trace][1][156vh788] Hybrid cpu=3.00%,41MB, cid=85,75, timer=62,0,639, clock=0,47,2,0,0,0,0,0,0, free=1, objs=(pkt:1612,raw:758,fua:851,msg:1774,oth:2,buf:50)
[2021-10-01 06:03:11.233][Trace][1][156vh788] **RTC: Server conns=13,** rpkts=(69,rtp:0,stun:5,rtcp:64), spkts=(1528,rtp:1523,stun:5,rtcp:0), rtcp=(pli:4,twcc:0,rr:0), rnk=(1,1,h:26,m:0), loss=(r:0,s:538), fid=(id:0,fid:69,ffid:0,addr:5,faddr:69)
```
1. SRS的配置如下(Config):
```
listen              1935;
max_connections     3;
daemon              off;
http_api {
    enabled         on;
    listen          1985;
}
http_server {
    enabled         on;
    listen          8080;
}
rtc_server {
enabled on;
# Listen at udp://8000
listen 8000;
#
# The $CANDIDATE means fetch from env, if not configed, use * as default.
#
# The * means retrieving server IP automatically, from all network interfaces,
# @see https://github.com/ossrs/srs/wiki/v4_CN_RTCWiki#config-candidate
candidate *;
}
vhost __defaultVhost__ {
#   cluster {
#        origin_cluster  on;
#        coworkers       127.0.0.1:1985;
#    }
    http_remux {
        enabled     on;
    }
    hls {
        enabled         off;
    }
    rtc {
      enabled     on;
      rtc_to_rtmp on;
    }
}
```

**重现(Replay)**

> 重现Bug的步骤(How to replay bug?)

1. `xxxxxx`
1. `xxxxxx`
1. `xxxxxx`

**期望行为(Expect)**

> 描述你期望发生的事情(Please describe your expectation)

**Comment by winlinvip** [comment]
I've looked through your questions, and it seems like you are expecting the following: For WebRTC push-pull streaming, you want to use `max_connections` to limit the total number of connections, right?

1. This feature has not been implemented yet, but it should be done. It is used to limit the system connection count, which is generally for preventing system overload, such as 1000 connections. If it exceeds this number, the connection will be rejected.
2. You set `max_connections 3;` to limit the number of connections. Is this for authentication purposes? If you want to verify whether each person has permission, don't use this. Instead, use the HTTP Callback, which has already been implemented.

**Comment by elin72215** [comment]
> 我看了下，你期望的应该是：WebRTC推拉流，用`max_connections`限制总数目，是吧？
> 
> 1. 这个功能还没实现，应该要做。它是限制的系统连接数，一般是为了防止系统超载的，比如1000个，超过这个就拒绝连接了。
> 2. 你设置为`max_connections 3;`，限制这么小是用来鉴权的么？如果要验证每个人是否有权限，不要用这个，应该用HTTP Callback，已经实现了。

        您好，我設置`max_connections 3;`是為了測試超過連接數會發生什麼事，剛好測到WebRTC推拉流不受該變數節制而已，我主要是想知道SRS的WebRTC推拉流架構來判斷是否使用其作為直播使用，假設直播主A用WebRTC推流，客戶端B用WebRTC拉流：
1. SRS是充當信令服務器只是協助兩方建立點對點連線，B直接從A拉流？
2. 還是實際上是A先跟SRS建立點對點連線推流，SRS接收流之後，SRS跟B建立點對點連線由B拉流？

**Comment by elin72215** [comment]
> 我看了下，你期望的应该是：WebRTC推拉流，用`max_connections`限制总数目，是吧？
> 
> 1. 这个功能还没实现，应该要做。它是限制的系统连接数，一般是为了防止系统超载的，比如1000个，超过这个就拒绝连接了。
> 2. 你设置为`max_connections 3;`，限制这么小是用来鉴权的么？如果要验证每个人是否有权限，不要用这个，应该用HTTP Callback，已经实现了。

        您好，我設置 max_connections 3; 是為了測試超過連接數會發生什麼事，剛好測到WebRTC推拉流不受該變數節制而已，我主要是想知道SRS的WebRTC推拉流架構來判斷是否使用其作為直播使用，假設直播主A用WebRTC推流，客戶端B用WebRTC拉流：
1. SRS是充當信令服務器只是協助兩方建立點對點連線，B直接從A拉流？
2. 還是實際上是A先跟SRS建立點對點連線推流，SRS接收流之後，SRS跟B建立點對點連線由B拉流？

**Comment by elin72215** [comment]
您好，我設置 max_connections 3; 是為了測試超過連接數會發生什麼事，剛好測到WebRTC推拉流不受該變數節制而已，我主要是想知道SRS的WebRTC推拉流架構來判斷是否使用其作為直播使用，假設直播主A用WebRTC推流，客戶端B用WebRTC拉流：
1. SRS是充當信令服務器只是協助兩方建立點對點連線，B直接從A拉流？
2. 還是實際上是A先跟SRS建立點對點連線推流，SRS接收流之後，SRS跟B建立點對點連線由B拉流？

**Comment by elin72215** [comment]
Hello, I set `max_connections 3` to test what would happen when the connection limit is exceeded. I just happened to find that WebRTC push-pull streaming is not restricted by this variable. My main concern is to understand the architecture of SRS's WebRTC push-pull streaming to determine whether to use it for live streaming. Suppose live streamer A uses WebRTC to push the stream, and client B uses WebRTC to pull the stream:

1. Does SRS act as a signaling server only to assist both parties in establishing a peer-to-peer connection, with B directly pulling the stream from A?
2. Or is it that A first establishes a peer-to-peer connection with SRS to push the stream, and after SRS receives the stream, SRS establishes a peer-to-peer connection with B for B to pull the stream?

In other words, are the peer-to-peer connection targets between the two users each other or SRS?

**Comment by duiniuluantanqin** [comment]
1. SRS uses an SFU architecture, and all streams are relayed through SRS. So, the process is that A pushes the stream to SRS, and B pulls the stream from SRS.
2. SRS did not support WebRTC before and was an RTMP server. `max_connections` was used to limit the number of RTMP connections. Since WebRTC uses UDP transmission, do we need to limit the number of connections?

**Comment by elin72215** [comment]
So, to implement WebRTC live streaming with WebRTC push streaming and a large number of WebRTC pull streams (for example, hundreds of thousands or millions of clients), as long as there are not too many push streams, a single SRS server can handle it? If we need to expand the push streaming scale, we can use the Origin Cluster mode, right? On the other hand, using RTMP push streaming and WebRTC pull streaming with a cluster would be limited because the transmitted stream remains RTMP until it actually goes out?

**Comment by duiniuluantanqin** [comment]
Yes, I talked with Mr. Yang yesterday, and this should be done. You are welcome to submit a PR (pull request).
